crates/matrix-sdk/src/http_client.rs

```diff
@@ -221,32 +206,90 @@ impl HttpClient {
                     }
                 };
 
-                let raw_response = self
+                let response = self
                     .inner
                     .send_request(clone_request(&request), config.timeout)
                     .await
                     .map_err(error_type)?;
 
-                trace!("Got response: {raw_response:?}");
-
-                let response = Request::IncomingResponse::try_from_http_response(raw_response)
-                    .map_err(|e| error_type(HttpError::from(e)))?;
-
                 Ok(response)
             };
 
             retry::<_, HttpError, _, _, _>(backoff, send_request).await?
         };
 
         #[cfg(target_arch = "wasm32")]
-        let response = {
-            let raw_response = self.inner.send_request(request, config.timeout).await?;
-            trace!("Got response: {raw_response:?}");
+        let response = self.inner.send_request(request, config.timeout).await?;
 
-            Request::IncomingResponse::try_from_http_response(raw_response)?
+        Ok(response)
+    }
+
+    #[instrument(
+        skip(self, access_token, config, request, user_id),
+        fields(config, path, user_id, status)
+    )]
+    pub async fn send<R>(
+        &self,
+        request: R,
+        config: Option<RequestConfig>,
+        homeserver: String,
+        access_token: Option<&str>,
+        user_id: Option<&UserId>,
+        server_versions: &[MatrixVersion],
+    ) -> Result<R::IncomingResponse, HttpError>
+    where
+        R: OutgoingRequest + Debug,
+        HttpError: From<FromHttpResponseError<R::EndpointError>>,
+    {
+        let span = tracing::Span::current();
+
+        let config = match config {
+            Some(config) => config,
+            None => self.request_config,
         };
 
-        Ok(response)
+        // Avoid logging the `None`.
+        span.record("config", debug(config));
+
+        // The user ID is only used if we're an app-service. Only log the user_id if
+        // it's `Some` and if assert_identity is set.
+        if config.assert_identity {
+            span.record("user_id", user_id.map(|u| u.as_str()));
+        }
+
+        let auth_scheme = R::METADATA.authentication;
+        if !matches!(auth_scheme, AuthScheme::AccessToken | AuthScheme::None) {
+            return Err(HttpError::NotClientRequest);
+        }
+
+        trace!(request_type = type_name::<R>(), "Serializing request");
```

Comment:
I think it would make sense to move this `trace` into `serialize_request` (and _maybe_ use `#[instrument(skip_all, fields(request_type = type_name::<R>()))]` instead of the `trace!` argument).

Reply:
I wanted to avoid adding another span just for this, since it adds the name of the span to the logs, which adds noise. I don't think it's useful to know that this line gets called in `serialize_request()`.

Reply:
Yeah, doesn't necessarily need a new span. I think actually Ruma might already log this anyways. But I think it still makes sense to move the `trace!` into `serialize_request`.